### ISSUE TITLE
Support AES-GCM

### ISSUE THREAD

**Issue by FranklinWhale** [dotnet/corefx#7023]
Currently I am using AuthenticatedAesCng in [Security.Cryptography.dll](https://clrsecurity.codeplex.com/) for AES-GCM encryption and decryption in a .NET application. I would like to have it added to CoreFX.

**Comment by bartonjs** [comment]
@FranklinWhale Oooh, I see, are you taking me up on my request to request missing functionality here? If so, that's the first evidence anyone has read my "Did you know?" from the codeplex project :).

It almost certainly won't look like it does in Security.Cryptography, but we are looking into adding support for AEAD modes.

If there's particular usage with that API that you adore, or hate, feel encouraged to speak up.

**Comment by FranklinWhale** [comment]
Yes, I did read the notice :)

I think the current way of using AuthenticatedAesCng is a bit too complicated. I hope there will be static methods available. For example:

```cs
public static class AesAead {
	public static class Gcm {
		public static AuthenticatedCiphertext Encrypt(byte[] key, byte[] iv, byte[] plaintext, byte[] authenticatedData = new byte[0]);
		
		// Take a stream of the plaintext
		// Return a stream of the ciphertext followed by an authentication tag
		public static Stream Encrypt(byte[] key, byte[] iv, Stream plaintext, byte[] authenticatedData = new byte[0]);
		
		public static byte[] Decrypt(byte[] key, byte[] iv, AuthenticatedCiphertext authenticatedCiphertext, byte[] authenticatedData = new byte[0]);
		
		// Take a stream of the ciphertext followed by an authentication tag
		// Return a stream of the plaintext
		public static Stream Decrypt(byte[] key, byte[] iv, Stream ciphertext, byte[] authenticatedData = new byte[0]);
	}
}

public class AuthenticatedCiphertext {
	public byte[] Ciphertext {
		get;
		set;
	}
	public byte[] AuthenticationTag {
		get;
		set;
	}
}
```

**Comment by FranklinWhale** [comment]
Modified example above

**Comment by drbo** [comment]
I think that there need to be high-level wrappers that hide nonce management from the user.  Nonce reuse is disastrous in AES-GCM.

**Comment by mmaguigan** [comment]
While I am not against having easy to use static methods they should be the after thought. I think we need to be thinking in terms of streams, and correspondingly, support for async-await (if the stream we're writing to supports async-await). Using byte arrays for symmetric cryptography encourages developers to load blobs entirely into RAM before encrypting or decrypting, potentially denial of service risk in cases where payloads can measure in the gigabytes rather than simple messages.

**Comment by DemiMarie** [comment]
That works for encryption, but it is a no go for decryption: for decryption
you need to load all data into memory before decrypting to avoid releasing
unverified plaintext (which an attacker can use to decrypt messages of
their choosing).  That is why libsodium (the best example I know of of a
good crypto library; .NET could do much worse that just provide libsodium
bindings) does not support streaming AEAD.  Streaming AEAD requires a
higher-level protocol with intermediate authentication tags.  There is a
great .NET library for this.

On Jul 20, 2016 11:18 AM, "mmaguigan" <notifications@github.com> wrote:

While I am not against having easy to use static methods they should be the
after thought. I think we need to be thinking in terms of streams, and
correspondingly, support for async-await (if the stream we're writing to
supports async-await). Using byte arrays for symmetric cryptography
encourages developers to load blobs entirely into RAM before encrypting or
decrypting, potentially denial of service risk in cases where payloads can
measure in the gigabytes rather than simple messages.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/dotnet/corefx/issues/7023#issuecomment-233982047>, or mute
the thread

**Comment by mmaguigan** [comment]
I think we want to be careful of prescribing high level constraints on low level APIs. In abstraction layers that's probably fine, but I think the primitives shouldn't concern themselves with protecting the user from shooting themselves in the foot. If the user of a low level API doesn't know any better and emits unauthenticated plain text to a stream expecting trusted content that's on them - just as we would let an unwitting developer make improper use of ECB if they really want too. A high level abstraction can be offered to protect them from making these sorts of mistakes... and emitting unauthenticated plain text vs. preventing a denial of service will probably take some clever thinking.

**Comment by DemiMarie** [comment]
The problem is that releasing unauthenticated plaintext is a vulnerability
if it gets to an attacker, even if it never gets to a party that expected
it to be verified.  I suspect that you could recover the authentication key
in realistic scenarios.

Fortunately, a known solution exists: break the message into multiple
smaller messages with distinct, linearly-increasing nonces.  AES-GCM needs
such a mechanism anyway for messages that are too long.

**Comment by mmaguigan** [comment]
I am saying that the implementation of straight AES-GCM should be pure and to the standard, embedding higher level protocols in a low level API to fix shortcomings of GCM don't belong at this level; this is critical for compatibility with other implementations (in other languages) and those who will forced to reinvent a very dangerous wheel because their hands are tied by the primitives and they cannot integrate with other systems otherwise.

It is an exercise for the higher level abstractions (or the engineer using block primitives directly) to deal with deciding on non-primitive approaches and what not, and at that level I am totally on board and excited to see what comes :)

**Comment by DemiMarie** [comment]
I agree.  However, the default method should just use byte arrays (simple
and safe).

**Comment by mmaguigan** [comment]
I have no problem with supporting byte arrays via the standard Init, Update, Final (IUF) pattern or whatever the community/microsoft comes up with - I'm sure it will be great. I do however have a problem shoehorning every implementation through simplistic methods like ```Encrypt(byte[] plaintext)```, it's cute but it doesn't scale. Whereas streams, IUF patterns and like methods scale well and work for both simplistic needs and high scale enterprise needs. I trust everyone understands this well enough and will make sure everyone's use cases are satisfied as well as can be hoped for.

exciting times! :smiley:

**Comment by DemiMarie** [comment]
@mmaguigan The problem is that, _with AES-GCM and ChaCha20-Poly1305_ (the latter of which also needs to be included in .NET Core), releasing unverified plaintext allows the attacker to decrypt messages of their choosing:

1. Find a message M that you want decrypted.
2. Create a message B that has the same nonce as M.  Send it.
3. The vulnerable target releases the unverified output P of the decryption (say having concluded that it is irrelevant, since it contains garbage).
4. The attacker XORs B, M, and P and has the decryption of M.

I know that simplistic methods like `Decrypt(byte[] plaintext)` don't scale.  My proposal is to provide such methods for use with small amounts of data, and something like [StreamCryptor] for larger amounts of data.  StreamCryptor is _designed_ to handle incremental authenticated encryption, and is immune to the vulnerability I mentioned.  It supports asynchronous encryption too.

[StreamCryptor]: /bitbeans/StreamCryptor

**Comment by mmaguigan** [comment]
I understand your point entirely. My point is that this is the wrong place to place safety-mechanisms - this is a primitive that **will** be relied upon for interoperability - and if you don't provide this primitive as per the standard (without forcing the user to load the entire cipher text into RAM - which is in and of itself a terrible practice with data sets of an unknown size) they will be forced into implementing the primitive themselves which could be disastrous.

Please remember that properly writing the plaintext to the user provided stream is **not** the equivelent of leaking plaintext in an AEAD system. Nobody said the stream the developer is giving you is writing to a consumer directly expecting trusted data. They could be reading a GCM protected stream from another node in an enterprise distributed file server and writing the content to disk (possibly encrypting it on the fly with the devices own DKEK or other KEK). They can happily abort the operation and delete the file once the tag is compromised without any exposed security vulnerability. This attack is only an attack when the vulnerable target leaks the plaintext, which no one is describing here.

The primitive needs to be *primitive*. Higher level abstractions are free to place high level constraints on consumers, but this needs to be done to the letter for compatibility and to prevent users from inventing their own crypto (or relying on Jacks Super Awesome NuGet Crypto library). You can't start telling users the reason Decrypt(byte[]) doesn't scale is because you shouldn't be using it, use my Hybrid AES-GCM and throw away all of the other systems you own that don't speak it.

Sidebar, I agree, perhaps the title should be changed to incorporate more related algorithms.

**Comment by DemiMarie** [comment]
Ah okay.  That makes sense.  I can easily imagine a system that decrypts to
a temporary file in a protected directory also used for other secrets, and
only makes it accessible to the rest of the system if the tag checked out.
Such a system would be secure.

I do think that the API documentation should come with stern warnings
explaining exactly what can go wrong, how it goes wrong (the actual
exploit), that the information must not be exposed if their is a connection
abort, and encouraging use of a higher-level API where possible.  TLS,
incidentally, does chunking internally, as does SSH.

On Jul 21, 2016 10:34 PM, "mmaguigan" <notifications@github.com> wrote:

I understand your point entirely. My point is that this is the wrong place
to place safety-mechanisms - this is a primitive that *will* be relied upon
for interoperability - and if you don't provide this primitive as per the
standard (without forcing the user to load the entire cipher text into RAM
- which is in and of itself a terrible practice with data sets of an
unknown size) they will be forced into implementing the primitive
themselves which could be disastrous.

Please remember that properly writing the plaintext to the user provided
stream is *not* the equivelent of leaking plaintext in an AEAD system.
Nobody said the stream the developer is giving you is writing to a consumer
directly expecting trusted data. They could be reading a GCM protected
stream from another node in an enterprise distributed file server and
writing the content to disk (possibly encrypting it on the fly with the
devices own DKEK or other KEK). They can happily abort the operation and
delete the file once the tag is compromised without any exposed security
vulnerability. This attack is only an attack when the vulnerable target
leaks the plaintext, which no one is describing here.

The primitive needs to be *primitive*. Higher level abstractions are free
to place high level constraints on consumers, but this needs to be done to
the letter for compatibility and to prevent users from inventing their own
crypto (or relying on Jacks Super Awesome NuGet Crypto library). You can't
start telling users the reason Decrypt(byte[]) doesn't scale is because you
shouldn't be using it, use my Hybrid AES-GCM and throw away all of the
other systems you own that don't speak it.

Sidebar, I agree, perhaps the title should be changed to incorporate more
related algorithms.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/dotnet/corefx/issues/7023#issuecomment-234439382>, or mute
the thread

**Comment by mmaguigan** [comment]
Totally agree. Hopefully with higher level abstractions and stern warnings most users will not be worrying themselves with proper implementation of AES-GCM (or AEAD in general maybe) and will rely on the framework to *do the right thing* (or TLS implementation) for their use case, but at least by making scaling basic primitives available we will have integrators and the engineers of advanced systems with less obvious use cases covered.

Good discussion @DemiMarie :smiley:

**Comment by DemiMarie** [comment]
You too @mmaguigan

**Comment by sdrapkin** [comment]
@mmaguigan @DemiMarie If you want to see .NET encryption done right &ndash; ex. streaming AEAD with full async-await support (nothing custom: it simply supports awaitable CryptoStream) which also does not leak unverified plaintext &ndash; look at [Inferno](http://SecurityDriven.NET/inferno/).

This is how corefx should've done it.

**Comment by DemiMarie** [comment]
I agree, but with more algorithms (ChaCha20-Poly1305 and Ed25519 for
starters).

**Comment by mmaguigan** [comment]
I am afraid I do not agree. That library appears to have the same problems @DemiMarie and I discussed earlier in the interoperability department. You can't roll your own flavor of chunking AEAD algorithms and expect interoperability. It also appears that the above mentioned library does not even use AES-GCM, which is the OP's subject of this issue.

I truly admire the effort and initiative put forth in the library to address the short comings of the framework (and common transform implementations), but interoperability with other languages is a critical concern in that no one will use your implementation (or they will drop it as soon as they realize...) if they cannot work with other implementations written in Java for example.

I agree with both of you on the matter of unauthenticated information disclosure, but we can't use the framework as a soap box for change, we need to voice our opinions to a wider audience.

**Comment by sdrapkin** [comment]
@mmaguigan I'm not advocating for copy-paste from Inferno into corefx &ndash; I'm merely suggesting the same approach that I've followed in Inferno, but with AES-GCM. I may be offtopic since this thread asks for AES-GCM (as you mentioned), but you yourself pointed out the issues with AES-GCM (ex. nonce reuse problems). Inferno essentially uses the same mechanisms as AES-GCM (same AES in CTR mode) but in a nice structure that eliminates many AES-GCM woes while keeping all the benefits of AES-CTR.

I also disagree with your compatibility comments. Inferno relies on AES-CTR (which is essentially AES-ECB with counters) and SHA384 hash. AES-ECB and SHA384 crypto primitives are available in virtually every platform and SDK, and will continue to be present for years to come. AES-GCM is a lot less available (case in point: .NET).

We all know that there are no chunking AEAD standards (other than, perhaps, TLS). I've made a sane, rational choices in Inferno (choice of crypto primitives, chunk size, tag size, format). Again, my point was not to advocate for copy/pasting Inferno into corefx, but to show a good way of implementing a streaming AEAD in .NET.

If you've seen how ASP.NET DataProtection logic works (similar ideas to Inferno) - it is also not standard but MS did it anyway. It's a clear MS precedent of going with a working approach vs. waiting for standardization.

**Comment by bartonjs** [comment]
Providing feedback on what shape you might like an API to go and commenting on the shape that someone else has proposed are both fine.  But this does not feel like an appropriate venue for product advertisement, or product critique.

I will say, though, that while we like the "bytes in, bytes out" approach; as a primitive we will probably stick with leaving things as individual components; such as the AuthenticationTag value being presented in a separate byte[] than the CipherText.  If there's a common IETF format then we can be persuaded to make accelerators for those formats; but when the primitives start doing data hiding it just results in multiple parallel endeavors for solving the same problem just to facilitate protocol interop.

**Comment by DemiMarie** [comment]
I think that the simple API should be the one that users are encouraged to
used, and that all other APIs should have warnings stating exactly what
must be done for safe use.  Note that a simple API is enough for TLS.

We DO NOT want to become like OpenSSL where most users don't use the
library correctly.

**Comment by karelz** [comment]
We need API proposal for each algorithm.
Keep in mind that each new algorithm has to receive Microsoft crypto-board approval - we can help get it started, but there is no guarantee about the result.

**Comment by sidshetye** [comment]
@bartonjs What's the status of having AES GCM in dot net standard? Also where is the dot net standard documentation for ECDH and ECDSA? Thanks

**Comment by bartonjs** [comment]
None of .NET { Core, Standard, Framework } have GCM or CCM.  It's still on the TODO list, and we still want to do it, but other things keep coming up.  The AE model of E(k,n,p) => (c,t) and D(k,n,p,c,t) => p doesn't work well with ICryptoTransform, and so we need to think up a new model for dealing with it.

Neither .NET Standard nor Core have ECDH, only .NET Framework.

ECDsa is part of .NET Standard, and its documentation lives at https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.ecdsa?view=netframework-4.7.  From a cert you'd use `cert.GetECDsaPublicKey()` or `cert.GetECDsaPrivateKey()`.

**Comment by sidshetye** [comment]
Thanks @bartonjs - we really need AES GCM primitives for Suite B plus NIST only recommends AE ciphers esp. GCM. On .NET Framework we can use the CLR Security layer to PInvoke (for both speed and FIPS usage) but this ought to be a first class citizen especially in the . What can we do to prioritize? Same goes for ECDH (secp256/384/521 curves).

Re: the new interface for AE ciphers, yes, that's a given. You've seen `AuthenticatedSymmetricAlgorithm` in CLR Security (e.g. https://blogs.msdn.microsoft.com/shawnfa/2009/03/17/authenticated-symmetric-encryption-in-net/) right? 

In summary: Great work here but what can we do to prioritize this effort?

**Comment by Drawaes** [comment]
What about breaking with tradition and going for something "lower level" that could then have a "pretty" array based wrapper put ontop if needed?

`public interface IBulkCipherKey : IDisposable
 {
        int TagSize { get; }
        int Update(ReadonlySpan<byte> input, Span<byte> output);
        int Update(Span<byte> inputAndOutput);
        int Finish(Span<byte> inputAndOutput);
        int Finish(ReadonlySpan<byte> input, Span<byte> output);
        void AddAdditionalInfo(ReadonlySpan<byte> additionalInfo);
        void GetTag(Span<byte> span);
        void SetTag(ReadOnlySpan<byte> tagSpan);
 }
`

**Comment by bartonjs** [comment]
@sidshetye Yep, I agree on all parts.  (Though Suite B is deprecated.  See http://web.archive.org/web/20150815072948/https://www.nsa.gov/ia/programs/suiteb_cryptography/index.shtml and https://www.iad.gov/iad/library/ia-guidance/ia-solutions-for-classified/algorithm-guidance/cnsa-suite-and-quantum-computing-faq.cfm -- iad.gov uses a non-public-trusted root... as mentioned on https://www.nsa.gov/what-we-do/information-assurance/)

ECDH is tracked by #8158.  GCM (and CCM) by this issue.

`AuthenticatedSymmetricAlgorithm` as-is leaves much to be desired.  For one, it was designed during an era of "properties are great, you configure things and then hit go".  But that makes for invalid states, or some questionable default values.

We need an API which

* Supports plain-AE and AEAD.
* Doesn't set you up for immediate failure with a lack of nonce management.
* Definitely freaks out when "TransformFinalBlock" hits a tag mismatch.
* Ideally avoids looking like a streaming API when it isn't.  (e.g. CLRSecurity's CCM buffers; I forget if that's "because CCM" or "because CCM needs a length up-front", or what)

Certainly, something one-shotty works:

```C#
public static class AuthenticatedAes
{
    public static [something] EncryptGCM([something] key, [something] nonce, ReadOnlySpan<byte> input, ReadOnlySpan<byte> authenticatedData, Span<byte> output, Span<byte> tagOutput);
    public static [something] DecryptGCM([something] key, [something] nonce, ReadOnlySpan<byte> tagOutput, ReadOnlySpan<byte> input, ReadOnlySpan<byte> authenticatedData, Span<byte> output);

    public static [something] EncryptCCM([something] key, [something] nonce, ReadOnlySpan<byte> input, ReadOnlySpan<byte> authenticatedData, Span<byte> output, Span<byte> tagOutput);
    public static [something] DecryptCCM([something] key, [something] nonce, ReadOnlySpan<byte> tagOutput, ReadOnlySpan<byte> input, ReadOnlySpan<byte> authenticatedData, Span<byte> output);
}
```

And so on for EAX, OCM, et cetera.  But

* How would we allow it to work with persisted keys?
    * Though I've been advised that "persisted key" + (GCM or CCM) == Bad Idea (nonce reuse concerns)
* Static methods make it harder to support crypto agility

So, yeah.  We know we need to do it, but it's about balancing this with other things that people are asking for.  Your input of saying you care helps, particularly since we're currently trying to figure out what the next tasks are.

**Comment by sdrapkin** [comment]
@sidshetye @bartonjs We use [Inferno](http://SecurityDriven.NET/inferno/)'s AEAD for full FIPS & Suite-B compliance (have been running multiple DoD solutions with Privacy Overlays and RMF ATO on .NET 4.7).
> The AE model of E(k,n,p) => (c,t) and D(k,n,p,c,t) => p doesn't work well with ICryptoTransform

Inferno implements AEAD via `ICryptoTransform`, which works great.

**Comment by Drawaes** [comment]
Static methods kinda suck if you write software that takes in "some crypto" algo and does stuff with it. You end up having to switch and wrap and all sorts of nasty stuff. Plus you have issues with isolating your code for testing.... 

As for persisted keys, its not something I have thought about but yeah its a bit of an issue considering windows CNG vs OpenSSL Crypto (and I presume SecureTransport) have very different handles. 

But that is why I wonder if having a class "instance" works better, the constructor for the above for AES-GCM for instance could take an IV/Nounce etc but throw an exception if its not "reset" between Finished and the next update.  

I also tend to agree with what I hinted at and you seem to carry on that, if changes are going to be made, Span/ReadonlySpan should be snuck in at the same time.

**Comment by jnm2** [comment]
[Android's AES/GCM API has `setRandomizedEncryptionRequired`](https://developer.android.com/reference/android/security/keystore/KeyGenParameterSpec.Builder.html#setRandomizedEncryptionRequired(boolean)) when creating a key which throws an exception if you try to provide your own IV when encrypting. By default it is turned on.

I am partial to that kind of approach.

**Comment by DemiMarie** [comment]
Wait, does it use random nonces for AES-GCM?????

NEVER use anything other than a counter for a CTR, GCM, or CCM nonce.  The
risk of collision is too high.  If you need to persist keys, generate a
session key by hashing some public coins with the key, and then use that
for the cipher.  Store the random bytes with the ciphertext.

**Comment by morganbr** [comment]
@jnm2 , while random nonces for GCM are an improvement, they're unfortunately only 96 bits. That effectively means that there's an unacceptably high chance that the same nonce will be chosen twice, which permanently breaks the security of the key.

**Comment by jnm2** [comment]
Good to know. It does in fact use random nonces. However, the same type of approach would work. Instead of randomized IVs, the API forces the IV to increment. The API would need to require a way to load and save the next IV for that key. I imagine the performance would be at least on par if not greater than simply taking a `byte[] iv` parameter.

**Comment by jnm2** [comment]
Yeah, the more I think about scenarios like parallel encryption of a series of individual messages, the more I realize that that doesn't belong in a low-level API.

**Comment by sidshetye** [comment]
static vs instance: not much of a preference as long as they're all thread safe x-platform. Either ways, somewhere there will be a switch statement setting either a param or calling the specific function or initializing a lamba.

Stream vs Byte[] vs whatever: I prefer the API stays low level rather than attempting to encapsulate logic that makes it inflexible. So I prefer Stream > byte[] >anything else. GCM decryption in theory is supposed to output a special `FAIL` symbol on tag mismatch but in practice, good luck with that approach if you're decrypting a multi-gigabyte stream. IMHO, better stream to the caller and throw an exception on tag mismatch. The caller can buffer the stream if they want to (resources allowing) and can also determine how they want to react to the exception. Crypto is full of (careful) performance optimizations for practicality and my $0.02 is that this is one of them.

IV: I think the important thing is that the API support a passed in IV so there in interop with other parts (e.g. IV comes from another subsystem). Having some helper logic to support auto-gen or auto-incremented IV is ok, but it should be a secondary goal (IMHO). Since we're talking about IV collisions, I think it's a bit far fetched. For a 2^96 number space - which is MASSIVE - collisions are about 1/2^48. That's 1 part in 39,614,081,257,132,168,796,771,975,168. I'd like to know who is encrypting so much data without ever rotating their AES encryption key ...

Should we break out the crypto discussion(s) separately (or take to crypto stack exchange) and keep this just for the project managament aspects of AES GCM/CTR in .net standard?
----
Sidenote: @bartonjs , yup, on suite B deprecation. But I do think it's mostly for EC given they are not quantum resistant (AES is). Till we get quantum resistant algos approved, we've got today's workflows.

**Comment by morganbr** [comment]
@sidshetye , the crypto discussion is relevant to the extent that it shapes the API -- we don't yet know how to make the API work well with the nonce management problems.

**Comment by Drawaes** [comment]
My thoughts on Nounces

1. Random has to absolutely be off the table. Never ever use random.
2. IV HAS to be user suppliable because otherwise it makes this API useless for interoperability. 
3. It should by default implement incrementing the IV so if someone uses it "out of the box" it will sort it out for them. This means you need state, which means static methods are off the table.
4. Static vs instance isn't the same thing at all. If you write an app to an interface and later you would rather use a different encryption then it's possible. With static you are putting in switch statements etc... If you want a code example I can provide one but static anything should be avoided especially in a core framework component. 

Is there any argument for static that I am not seeing?

**Comment by FranklinWhale** [comment]
Perhaps [openssl_encrypt](http://php.net/manual/en/function.openssl-encrypt.php) in PHP is a good reference. Almost all symmetric-key algorithms are supported via a single interface.

**Comment by Drawaes** [comment]
Terrible interface :p My issue with static methods is how to I encrypt split blocks/fragmented buffers. This is a huge requirement in networking especially because otherwise I need to ensure I gather the whole stream to send at once.

If we take a look at the underlying providers (I dunno about osx but openssl and schannel) there is pretty much a "context" that you can send multiple blocks to. Ideally the user doesn't want to have to maintain this state between calls or have to coalesce the data themselves.

https://github.com/Drawaes/Leto/blob/master/src/Leto.OpenSsl11/OpenSslBulkCipherKey.cs

Something better designed but along the idea of above would sort it out... and still allow all algos

**Comment by sdrapkin** [comment]
There is _no_ way to provide _safe_ AES-GCM APIs. Either you randomise the insufficiently-large 96bit IV-space, or you throw this at your users to correctly nonce the IV (which they won't). It's a pit of failure, either way.

@GrabYourPitchforks (MS) reached the same conclusion when designing ASP.NET [DataProtection](https://docs.microsoft.com/en-us/aspnet/core/security/data-protection/implementation/subkeyderivation) APIs which optionally do AESGCM. His approach was to do subkey derivation (which also happens to be similar to the approach used by [Inferno](http://SecurityDriven.NET/inferno/)). Why would MS ignore the insights of their own MS contributors with AESGCM experience?

D.J.Bernstein also created a 192bit-nonce xSalsa just to avoid having do deal with the 64bit-nonce of the original.

So the decision for MS seems to either:

- offer "standard" AESGCM, which will be misused by the majority of callers, and which would contribute to a ton of new insecure .NET-based code.
- offer "non-standard" AEAD with safe APIs that cannot be misused.
- do nothing (delay, procrastinate, wait for streaming-AEAD standardisation, etc).

DataProtection team and @GrabYourPitchforks chose the 2nd option.

**Comment by mmaguigan** [comment]
@sdrapkin,
We shouldn't be presumptuous about the skill of API users. This is why low-level and high-level API distinctions exist. As a side note, DJB's nacl allows the user to provide the nonce with [crypto_secretbox](https://nacl.cr.yp.to/secretbox.html) - likewise, libsodium also allows this with it's [AES-GCM](https://download.libsodium.org/doc/secret-key_cryptography/aes-256-gcm.html) support.

If you are re-using the same symmetric key with AES-GCM, you may randomly provision the nonce at key generation time and increment this random nonce with every re-use of the same symmetric key. I state this for @jnm2's mention of parallel encryption of individual messages with the same key.

Users need more than a simple encryption API (ala DPAPI), they need the ability to inter-operate with systems from other platforms *and* invent new technologies.

**Comment by sdrapkin** [comment]
@mmaguigan 
> Users need more than a simple encryption API (ala DPAPI), they need the ability to inter-operate with systems from other platforms and invent new technologies.

Users need simple, safe, AEAD encryption API **first and foremost**. .NET does not provide a simple AEAD encryption API - that IMHO must be addressed _before_ catering to skilled crypto-aware users (who, btw had AESGCM on .NET for a [long time](https://bitbeans.gitbooks.io/libsodium-net/content/advanced/aes256_gcm.html) via Libsodium, BouncyCastle, or custom implementations).

I agree that having both high-level APIs (for common scenarios) and low-level APIs (for sophisticated users and adhoc scenarios) would be great. But priorities matter.

**Comment by Drawaes** [comment]
I am all for a lower and higher level abstractions. I primarily care about the lower level ones and let others worry about the user friendly side.

For the lower level, ability to set/control the nounce. Ability to have a destructive or not encrypt, little to no allocations, ideally span over byte and finally fragmented buffers without any static would keep me quiet for a while. :)

**Comment by mmaguigan** [comment]
@sdrapkin, would you then agree that such an API would be constructed out of a set of low-level primitives? You have yourself a chicken and the egg problem. I'm afraid that what you are proposing is that such an API be kept ```internal``` and users be herded into simple API. This doesn't meet my use case as someone whose customers include large-scale enterprise (>15,000 empl.) - and I imagine it wouldn't meet the use case of most enterprise customers.

I agree with @Drawaes, the use of static methods seems a bit flip. I'm sure we can all come up with a good API that will satisfy the needs of the many. And then a good high-level API for the rest.

**Comment by sdrapkin** [comment]
@mmaguigan I'm not proposing to keep low-level APIs internal. As an example, `AES` itself is a low-level API with dozens of ways to misuse it (IV, chaining, padding, AE, etc, etc) - however it's a public API because it acts as a building block for other constructs. The .NET problem is that it **only** exposes unsafe APIs - there are no safe AEAD APIs. So poor users have no choice but to concoct some AES-encryption-scheme that seems good-enough (output looks like garbage). If you also offer them GCM in addition to CBC and ECB, you'd be removing 1 bad decision (lack of AE) while adding 2 more in addition to ~5 other bad decisions that are already there. A sophisticated user with a "need-raw-AESGCM" requirement will be satisfied (and I'm all for it), but the typical user (especially typical for a large-enterprise) would be stranded among a sea of bad choices -- that is unless a safe-to-use, secure, and simple API is available.

**Comment by Drawaes** [comment]
Sure but then they should be using something more akin to a Dataprotection style API no?

**Comment by sidshetye** [comment]
@Drawaes and other against random nonce - why? A 96bit crypto-random IV gives less than a 1/2^48 chance of collisions. NIST 800-38D section 8 allows for the security proof to stay as long as it's under 1/2^32 (we're FAR above that). 

> The probability that the authenticated encryption function ever will be invoked with the same IV and the same key on two (or more) distinct sets of input data shall be no greater than 2^-32

Even the original AES GCM paper by McGrew and Viega, in section 2.1 page 3 says:

> The primary purpose of the IV is to be a nonce, that is, to be distinct for each invocation of the encryption operation for a fixed key. It is acceptable for the IV to be generated randomly, as long as the distinctness of the IV values for each key is highly likely

So a crypto-random is certainly a practical and pragmatic engineering solution. I'm not convinced this adds to a "sea of bad choices". Still, I worry we get distracted talking about the higher level safe API while we have absolutely nothing in AE today.

My suggestions for priorities would be 

1. Finalize the low level AE interface (here; lets enumerate the open choices and lock each down)
2. Implement, test and release that
3. Finalize a high level "safe" AE interface (another ticket?)
4. Implement, test and release that

This issue's been open for over 1 year 3 months, so I'm worried a long drawn out discussion on the higher safe APIs or other stuff may further delay the basic low level AES GCM that's fast, flexible, x-platform and FIPS powered (where possible).

**Comment by Drawaes** [comment]
Because of the birthday problem.... You clash calculation is assuming a single random vector vs another
[Wiki Birthday Problem](https://en.m.wikipedia.org/wiki/Birthday_problem)

**Comment by jnm2** [comment]
Yep. Now I'm curious. What is the maximum number of times you can use a random IV before the probability of a birthday collision is too great?

**Comment by Drawaes** [comment]
So here is the thing with AES-GCM yoù shouldn't encrypt large amounts of data with a single tag... so you need to break it up... therefore you end up with a lot of random nonces. It's time for me to read the kids bedtime stories bit feel free to do the math.

**Comment by Drawaes** [comment]
Actually while we are at it... how do you set the nonce in the other direction or retrieve it and transmit it? What's the perf hit for a truly random number...

I could see a random starting point with a counter from there including wrapping... or as tls 1.3 does which is a random 12 byte number xor with the sequence number

**Comment by mmaguigan** [comment]
I do not want to postulate on the circumstances wherein a user would be using the same key over and over again for millions of operations without re-keying - I am certain there are legitimate protocols in existence. That said, that circumstance is probably seeded from two possible events; event A) a mistake, wherein the user hard coded an encryption key or is reading a static symmetric key from disk and using it for a **long** time, or event B) the user is implementing a legitimate protocol wherein the volume of traffic to re-keying rate is inadequately tuned.

We cannot (shouldn't) prevent the user in event A from inventing themselves a footgun; this user should consider a higher level API (which I agree should exist... AFTER we have a complete and beautiful low-level API). The user in event B should have the foresight to increment their nonce.

I don't think the goal of a quality API is to prevent users from shooting themselves in the foot, it's to create powerful tools. Some of those tools are indeed abstractions. But what .NET lacks is a decent, and complete suite of cryptographic primitives - I'll echo the need for non-government sponsored ciphers as well (eg. some of DJB's fine work). But I agree with @sidshetye, the work has to start - and at the bottom is where it begins.

It's reasonable to instantiate an AEAD cryptographic primitive with a random key and nonce. Between the key space and the nonce space, there is little risk there. If the user is making the mistake of setting the key, and leaving the random nonce - that has to be on the user. Perhaps an intellisense comment on the key is a reasonable middle ground.

@Drawaes, absolutely agree on incrementing randomly seeded nonce - that's the 80/20 middle ground to aim for. That said, I don't know that the low level API would have state from operation to operation to know whether or not to increment the nonce. Perhaps if ```Init()``` isn't called, increment the nonce between each call to ```Encrypt()```; if ```Init()``` is called, reset the key and nonce to a random value and the user can decide whether or not to change it to another value (or even offer an overload of ```Init()``` that allows you to pass a key and nonce (together!).

**Comment by Drawaes** [comment]
agreed... that was my
>It should by default implement incrementing the IV so if someone uses it "out of the box" it will sort it out for them. This means you need state, which means static methods are off the table.

So if I want to set it I can, if I don't it starts at random and increments.

**Comment by jnm2** [comment]
Using the square approximation, because the exact calculation is beyond Wolfram Alpha, you'd have to use about  6,074,001,000 random 96-bit nonces for the probability of at least two colliding to become greater than 2<sup>-32</sup>.

**Comment by jnm2** [comment]
You'd have to generate enough messages to overflow an `int` twice and almost a third time if you counted them.

**Comment by Drawaes** [comment]
Hahaha, math eh, get you everytime. Fair enough then use randoms for the higher level or "non" supplied, but it's not my use case anyway. Everything I have had to use had a predefined nonce generation method in it's spec, so as long as I don't have to pay the price of calculating a decent random every time just to throw it away. I am happy. :)

**Comment by Drawaes** [comment]
FYI 
86400 seconds per day 1000 requests per second, ~4b per 50 days (the number is closer to 4 than 6b) so if you didn't rotate your keys that often on a busy service you hit the 2^32 limit above... not that I think that is that viable, just saying :) Also you are eating entropy for your random numbers when you don't need them other than a starting point (if at all). FYI for TLS 1.2 OpenSSL uses a random starting point and increments so there is precedent.

**Comment by jnm2** [comment]
> the number is closer to 4 than 6b

I'm interested... did you find the number? I want to hang onto it.

**Comment by Drawaes** [comment]
I hate to invoke NIST but, it states not that the odds need to be 2^32 but that you can ONLY invoke the algo 2^32 times unless you use deterministic construction with 96bit IV's ..... section 8.3,

So that is only 4b roughly (what's 200m between friends) as the IETF documents point out, if you were say encrypting disk data, with 512 byte block sizes (to match a lot of hard disks) that is only 2tb of data, not a small amount but not impossible in todays world, and with the explosion of data really not that much of a stretch at all, seeing as I have a 1tb Samsung portable ssd using AES sitting in front of me.

[NIST GCM](http://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf)
>The following requirement applies to all implementations that use either 1) the deterministic construction with IVs whose length is not 96, or 2) the RBG-based construction, for IVs of any length.  In other words, unless an implementation only uses 96-bit IVs that are generated by the deterministic construction: 
 
>The total number of invocations of the authenticated encryption function shall not exceed 2^32, including all IV lengths and all instances of the authenticated encryption function with the given key.

**Comment by Drawaes** [comment]
I am now wondering what happens when my ssd runs out of IV's.... ?

**Comment by sidshetye** [comment]
Got a break and seeing github alerts :) .. 

I've confirmed Joseph's numbers and we're going fairly deep into the IV specifics. While on it, two threat models:

1. Attacker wants to decrypt *this* piece of encrypted data => requires 1/2^48 attempts on randomized IV/nonce => impossibly remote => we're ok

2. Attacker wants to decrypt *any* two pieces of encrypted data => birthday attack => 
    2a) requires 6 billion attempts onto your service to THEN have a one in 4 billion chance of IV reuse => remote => we're ok 
    2b) requires access to almost the data every encryped by that key to THEN have a one in 4 billion chance of IV reuse => remote => we're ok
    2c) 2a and 2b are two distinct, remote chances. Lets distill it into a single dimension by looking at the alf-half / 50% chance. A random IV requires 331,411,458,666,436 attempts (or access to that many encrypted blobs)

I do consider #1 a more practical threat model ("wants this high value data") rather than #2 ("gimme anything"), but regardless outcomes are remote. I'm at Crypteron, and we make key rotations just one click away, so I'm all for proper key rotations. I'm just saying that practically the random IV issue is very remote issue. I'm more worried about not having AES GCM itself as a first class citizen for our .NET line of products. I mean, we're not married to any particular crypto library but would rather use the language standards if it's x-platform, uses AES-NI and is FIPS certified.

----
PS: An alternative non-math argument - if this IV stuff were a threat, why would the designers of GCM and NIST allow for random IVs :) ?

**Comment by Drawaes** [comment]
They do, they allow 2^32 randoms per key, and as said if you are using my harddrive example above, even if the chance is remote, you are no longer NIST certified at that 2^32 region.

**Comment by DemiMarie** [comment]
That is why we need subkey derivation: use a hash (such as SHA512) to
generate the key and nonce from a master key and an arbitrarily long
nonce.  Then increment the nonce for subsequent messages.

This is the only safe construction.

**Comment by morganbr** [comment]
Subkey derivation is a fine solution, but it's a custom protocol, not AES-GCM. Doing things like that make the API more like DPAPI. I'm all for having an "opinionated" AEAD API that incorporates things like that (and/or safer algorithms than GCM), but that doesn't help folks who need to conform to a protocol.

**Comment by Drawaes** [comment]
@morganbr 100% agree which is why I think the auto IV in any form is a waste of time. If we look at it like this

1. You are implementing a protocol / scheme/ standard (your own or otherwise) . Then you probably know what you are doing or whoever wrote the protocol will ( we hope). Then nonce generation should be part of that and you will need your own scheme.
2. You just want a way to keep your super secret secrets, super secret :) then a data protection like api is fine. And prescriptive nonces and key generation etc are what you want and expect.

**Comment by vcsjones** [comment]
I'm curious how AES-GCM will be implemented on the Apple platform with CommonCrypto. While macOS has support for AES-GCM, it isn't publicly documented. You have to sneak in a header and use, technically, private APIs for AES-GCM/CCM on macOS.

**Comment by bartonjs** [comment]
@vcsjones Probably PlatformNotSupportedException.  (Though possibly via OpenSSL lightup).

**Comment by vcsjones** [comment]
>(Though possibly via OpenSSL lightup)

If that is on the table, I would strongly consider it.

**Comment by Drawaes** [comment]
If that is so, then will you need to ship a version on OpenSSL for OSX, I thought the old one with Macs doesn't support GCM? I could be wrong.

**Comment by sidshetye** [comment]
@bartonjs 'supporting' it via a PlatformNotSupportedException is really strange :/  ... my vote would be for the OpenSSL route too, especially if you're doing something similar for Linux. Could also allow for a FIPS-140 validated route through OpenSSL FIPS modules. As well as AES-NI and GCM operation acceleration that's in OpenSSL.

**Comment by bartonjs** [comment]
@dawaes We only work with OpenSSL 1.0.x, so we ignore the 0.9.8 version that ships with the OS.  By "Lightup" I mean "if OpenSSL 1.0.x is installed we'll use it for GCM, otherwise we will throw PNSE".

The only reason it's contentious is I just spent 9 months making it so we had no OpenSSL dependencies on a mac :smile:.

**Comment by sidshetye** [comment]
> The only reason it's contentious is I just spent 9 months making it so we had no OpenSSL dependencies on a mac 😄.

Could you share why? You've spent some time with it, so there might be something I (or the community) could learn by sharing that experience. BTW, another +1 for OpenSSL is that you'd get other modes like CCM etc 'for free'.

**Comment by vcsjones** [comment]
>I thought the old one with Macs doesn't support GCM?

Correct. macOS ships OpenSSL 0.9.8zh 14 Jan 2016 out of the box.

There are a few options here.

1. Just don't support AES-GCM on a Mac.
1. Use OpenSSL. Do a runtime check for support and a correct version of OpenSSL, if it's there, use it. 
1. If it's on the table, use LibreSSL which will be shipping in 10.13 out of the box. Apple is going to start shipping and updating LibreSSL in macOS to address the ancient version of OpenSSL they include out of the box. This will get out-of-the-box support in 10.13.
1. Use the "private" APIs in macOS to get AES-GCM support. An example of that is [here](https://gist.github.com/vcsjones/48f78cccd3c421b138865ff72ee0f84a). While it is undocumented, the headers that make it possible *are* released in Apple's OSS for CommonCrypto: https://opensource.apple.com/source/CommonCrypto/CommonCrypto-60027/Source/CommonCryptoSPI/CommonCryptorSPI.h.auto.html

**Comment by Drawaes** [comment]
@bartonjs I was also thinking that :) ahh the life of a dev ...

**Comment by vcsjones** [comment]
>Could you share why? You've spent some time with it, so there might be something I (or the community) could learn by sharing that experience.

I sympathize with the desire to be rid of OpenSSL on macOS. The one that comes out of the box is horribly out of date and Apple doesn't support it. That means CoreFX would need to ship OpenSSL with CoreFX, which is very undesirable because the impetus for patching it is now on the CoreFX team.

The other option is to just hope it's on the box. Homebrew is the most common way of doing this, but even homebrew struggles to support it. They cannot link it in to the common lib directories without breaking other software, so there is some manual creation of symlinks that needs to happen. It's a frustrating process.

**Comment by bartonjs** [comment]
> 1. Just don't support AES-GCM on a Mac.

This is the "if we can't solve the problem in a happy way".  If the platform doesn't support it, we'll report that.

> 2. Use OpenSSL. Do a runtime check for support and a correct version of OpenSSL, if it's there, use it.

This is likely.  `new RSAOpenSsl()` on macOS uses OpenSSL; but that's a very direct "I want to use OpenSSL" option.  For GCM/CCM it would be "see if we can do a safe-init, and instead of throwing the TypeInitializationException throw a PlatformNotSuppotedException".  It might be just as easy as

```C#
try
{
    CallSomeOtherTypeToDoOpenSSLGCM(args);
}
catch (TypeInitializationException e)
{
    throw new PlatformNotSupportedException(someMessage, e);
}
```

The problem is that it goes back to requiring a macPorts or Homebrew (or other) installation of OpenSSL 1.0.x; and that makes people sad when they try to build a standalone application (it's not standalone, since the destination system has a subtle prereq).

> 3. If it's on the table, use LibreSSL which will be shipping in 10.13 out of the box. 

It's a very not-preferred approach.  It means we'd be using 4 system libraries now, not 3.  And it would enforce a min-OS of 10.13, or still have to PNSE on 10.12.

> 4. Use the "private" APIs in macOS to get AES-GCM support. 

This is completely off the table.

**Comment by bartonjs** [comment]
> I sympathize with the desire to be rid of OpenSSL on macOS. The one that comes out of the box is horribly out of date and Apple doesn't support it.

Nor do we.  We won't call libcrypto.0.9.8.dylib, only libcrypto.1.0.0.dylib, which is not on-the box.

> That means CoreFX would need to ship OpenSSL with CoreFX, 

By policy we do not, and will not.

> The other option is to just hope it's on the box. Homebrew is the most common way of doing this

Yep, our official instructions for 1.x were to install OpenSSL via Homebrew, and do a minimal installation (linking the versioned libraries in, but not replacing the "unversioned" forwarder)

**Comment by sidshetye** [comment]
First Winter was coming
Now Winter is here
But AES GCM is nowhere near
With .NET Standard out of the gate
Can we have a clear release date?

**Comment by morganbr** [comment]
If AES GCM is what you seek
We need a good way to keep nonces unique
But don't let OSX support hold us back
Throw PNSE cuz it's not there on Mac

**Comment by lennybacon** [comment]
RNG might help on the nonce side

**Comment by sdrapkin** [comment]
My sense is that AES-GCM will not be available until a standard AE/AEAD API is decided upon (with AES-GCM acting as one implementation of that API).

Since no such standard AE/AEAD API is in the works, the choice made is (3):

> do nothing (delay, procrastinate, wait for streaming-AEAD standardisation, etc).

P.S. Suite-B is gone, and CNSA drops AES-GCM (intentionally). For interoperability there are other libs with AES-GCM implementations. AES-GCM is "legacy", just like 3DES - get used to it.

**Comment by mmaguigan** [comment]
@morganbr, a primitive should not enforce the means of keeping a nonce unique (despite FUD, this is indeed trivial) otherwise it would be impossible for .NET developers to maintain compatibility with systems written in other languages. A higher level abstraction should of course be provided, and for that, a method does exist, seed the nonce with a random value and increment from there on each message until re-keying (and then rinse and repeat). I worry that exposing primitives has been blocked by discussion around these higher level abstractions. Thankfully, P/Invoke keeps us from being high and dry (eg. p/invoke libsodium), but it's not optimal that the framework doesn't provide AEAD.

@sdrapkin, unless you know something I don't, AES-GCM is not a legacy block-cipher-mode. GCM is not the only block cipher mode that requires a unique nonce (or places reasonable constraints on the implementer for that matter), and it's not the cipher mode's fault if engineers do not adhere to it's safety constraints. AES-GCM has not been compromised as far as I am aware and shouldn't be compared to 3DES (sweet32 attack).

**Comment by Drawaes** [comment]
I completely agree with above, the primatives should not enforce constraints otherwise both ends of use HAS to be .net or changeable to match .net's System which is madness,.

And sorry to agree again but @sdrapkin CSA doesn't "drop AES-GCM" what it does is specify AES. It leaves the block mode up to the standard from my discussions and understanding of it. So say if TLS 1.3 says GCM then GCM it is, they have just stopped specifying the block/ahead mode completely for flexibility. Also note they also say in CNSA that AES-256 is the only cipher they are pretty sure will be safe in a Quantum computing world...

**Comment by bartonjs** [comment]
> Can we have a clear release date?

No, there are too many variables.  1) When release dates are for upcoming releases. 2) Whether or not we'll have an API decided on by the next release. 3) Whether or not we'll get an implementation in by that same release.

This issue is tagged as `api-needs-work`, meaning we have an understood goal, but not an API shape.  Looking back through the history here:

* https://github.com/dotnet/corefx/issues/7023#issuecomment-198458299
  * Proposed a static class AesGcm.
  * Some grumbling was had about staticity
* https://github.com/dotnet/corefx/issues/7023#issuecomment-311835795
  * Proposed using the API from the CLRSecurity project
  * That combines setup and finished in very awkward ways.
  * It also makes a whole new hierarchy such that Aes and AesGcm are fairly unrelated (which, maybe that's good).
* https://github.com/dotnet/corefx/issues/7023#issuecomment-311840928
  * Proposed a new low-level interface that could replace all of symmetric operations.
  * It combines setup, in-progress, and finished in not very .NETty ways.
* https://github.com/dotnet/corefx/issues/7023#issuecomment-311845550
  * Proposed static methods on Aes instead of a new static class.
  * Still grumblings about staticity
* https://github.com/dotnet/corefx/issues/7023#issuecomment-311846745
  * Asserted that ICryptoTransform is fine.
  * ICryptoTransform is doable, it's just not pleasant.  The caller needs to hang on to the value from CreateEncryptor, and when done to cast it to something else and query the tag output (or glue the tag on the end of the ciphertext).  So it doesn't create a pit of success.

Of the 83 (now 84) comments in this thread it looks like 5 have contributed API shape, and maybe that many have contributed commentary on the 5.

That said, trying to throw more brainpower at this and get it done for 2.1 is our current desire (see https://github.com/dotnet/designs/issues/11).  But that's not a commitment, and certainly not a date.  At this point rattling the cages won't help things get done any faster.

**Comment by sdrapkin** [comment]
@morganbr The main weakness of 3DES (Encryption-only model) is small block size, which is what sweet32 exploits - ie. _bruteforce_ is the attack. The exact same threat vector applies to GCM in AE model - ie. its main weakness is the small 96bit nonce space which cannot be randomised to maintain safety margins in same_key+high_volume applications.

Say, you've followed NIST and implemented 3DES with rekeying after 2^32 blocks (using a custom, "non-standard" rekeying logic, mind you). Great. And then NIST has on OMG moment and [reduces](https://beta.csrc.nist.gov/News/2017/Update-to-Current-Use-and-Deprecation-of-TDEA ) the rekeying requirement to 2^20.

Are we going to wait for the same to happen to AES-GCM under AE model? Ie. for a "sweet96" that leads NIST to another OMG moment and revised sp800-38d_sec8.3 (constraints on number of invocations)?

Let me borrow an argument-style from DJB's playbook. One cannot simultaneously believe that:

1. The use of algorithms and keys providing a security strength of 80 bits is **no longer approved** since it is no longer considered as providing adequate protection. [NIST SP800-175Bsec3.4].
2. AES-GCM with 96bit nonce space provides adequate protection _if only_ everyone were to rekey _early-enough_, for which _no standard exists_.

DJB had an [interesting analysis](https://blog.cr.yp.to/20151120-batchattacks.html) as well.

I never said that AES-GCM is a deprecated _mode_ - it is not, just like ECB is not a deprecated mode. Neither is very useful without constructs on top. I did say that AES-GCM should be considered "deprecated" as a sole AE primitive with nothing on top.

@Drawaes CNSA drops AES-GCM as one of the _only allowed_ AE modes (GCM/CCM), and instead makes the AE mode a _choice_, as long as Confidentiality+Authentication combo follows the guidelines. What are the guidelines? NIST SP800-175Bsec4.3. Either you do AE as 2 separate operations (ex. what [Inferno](http://securitydriven.net/inferno/) does), or you do it in a single operation (ex. GCM/CCM).

**Comment by sidshetye** [comment]
@sdrapkin your comment "_AES-GCM is "legacy" ... get used to it_." - can't say the industry agrees. Still, you're entitled to your opinion. If you don't need the API, don't use it. But then again, this is a ticket about crypto primitives, not higher level constructs or constraints, so I'd rather stay course. About alternatives: I don't know of any other API that does AES-GCM, with AES-NI, on .NET Framework and .NET Core (x-platform) and comes in FIPS-140-2 certified flavors (in some platforms). If you know, I'm seriously all ears. Lastly, I consider AE via two independent operations a step backwards compared to single steps like CCM or GCM (just my $0.02).

@bartonjs I ask to manage my roadmap risk - not to annoy. Without slack or email threads, GitHub comments are the best ways for communications. Having said that, would we get a faster release date if we were do deliver a keg of the team's favorite brew? For science and progress of course. Has worked in the past!

**Comment by sdrapkin** [comment]
@sidshetye People & Orgs much smarter than all of us (ie. "the industry") have built AES-GCM into TLS 1.2, which is what you're most likely using right now to read this comment.

And then the [NDA](https://github.com/nonce-disrespect/nonce-disrespect) showed that 64 bits of randomness (which is what's left in TLS 1.2 for GCM, since 32 bits are fixed per-connection), are inadequate. So in TLS 1.3 "the industry" has  _given up_ on GCM IV randomization, and will switch to record numbers xor'ed into IV.

The take-away is that AES-GCM with 96bit nonces was extremely difficult to get right for the smartest people in this field, and TLS 1.3 is still neither finished, nor widely deployed. The "hack" that made GCM work in TLS 1.2 and 1.3 is taking advantage of the fact that TLS just happens to be an online protocol with natural sequential non-repeating counters.

For offline protocols without obvious natural nonces, early-enough rekeying is the only way to use GCM safely (for which no standard exists).

**Comment by karelz** [comment]
@sidshetye The priority is pretty high on our Security roadmap - see https://github.com/dotnet/designs/issues/11.
We currently plan to land this feature in 2.1 (with [95% confidence](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/issue-guide.md#triage-rules---simplified)). See [official release dates](https://github.com/dotnet/core/blob/master/roadmap.md#ship-dates) for more details about timelines.

Given that this is a new feature & API, 2.1 is the earliest release in which we can deliver it.

In general there are 2 ways to influence when any feature is available:
1. Contribute
2. Influence our prioritization

However, as I mentioned, there is no earlier shipping vehicle than 2.1. So, contribution would make it appear earlier in previews, and our prioritization is already "right" from 2.1 perspective.

**Comment by Drawaes** [comment]
If we want to move this along we need a concrete API design. For that we need some agreement, 

So first issue I see is the static vs non static. The issue I see with static is that it's hard to hold state. You will need some sort of state struct or class anyway, especially if you are not encrypting all the data at once (say network or streaming).

Others thoughts for and against?

**Comment by morganbr** [comment]
I'm hoping to see a general-purpose AEAD API that allows us to expose other (less dangerous) types of AEAD and allows some degree of crypto-agility (switching algorithms based on configuration).

**Comment by mmaguigan** [comment]
@Drawaes, I would personally be sad to see us go with static. It would undoubtedly limit it's integration with new language features such as "extension everything". Likewise, depending on underlying platform, disposal is potentially valuable.

I'd like to see some of our nice new language features exposed here if possible (eg. spans).

**Comment by Drawaes** [comment]
Okay so, in general if we said we could move on from the static discussion. It really needs to be instance types? So that things can be extended etc.

The next question then would be .. and this one is not so easy nonce management.

I see two real use cases here and this is why the conflict is happening around this topic. You have the people that want to use this API for an existing standardised (or inhouse or whatever) protocol/encryption management etc. Then you have the people that want an API and don't have restrictions on how the nonce is generated.

So here is what I am suggesting (without a design for it yet, just a concept). We have a default of some description that lets you fall into the pit of success and auto generates a random nonce. We could even track how many times you have reset that object or whatever if the random nonce is in effect and throw if the number gets too high. But that could be turned off, then you would have to supply the nonce for each operation or it would be an exception.

This way the people that need to supply their own can, and for everyone else it just happens as close to a right way as possible.

Thoughts? (let's move this along so we can get to a solid design ;) ).

**Comment by jnm2** [comment]
Or the parameter could be called `dangerousNonce` with XML docs pointing you to the mentioned higher-level abstraction.

**Comment by Drawaes** [comment]
dangerous is off the table I would say, it already has an established meaning, to do with interior pointers and potential tearing etc

**Comment by jnm2** [comment]
```
System.​Span<T>.Dangerous​Create(Object, T, Int32)
System.​Read​Only​Span<T>.Dangerous​Create(Object, T, Int32)
System.​Span<T>.Dangerous​Get​Pinnable​Reference()	
System.​Read​Only​Span<T>.Dangerous​Get​Pinnable​Reference()
System.​Runtime.​Interop​Services.​Safe​Handle.​Dangerous​Add​Ref(Boolean)
System.​Runtime.​Interop​Services.​Safe​Handle.​Dangerous​Release()
System.​Runtime.​Interop​Services.​Safe​Handle.​Dangerous​Get​Handle()
System.​Net.​Http.​Http​Client​Handler.​Dangerous​Accept​Any​Server​Certificate​Validator
```

**Comment by mmaguigan** [comment]
I would say that the primitive API should simply do what it's told with the parameters it has set (eg. Key = X, Nonce = N), with a default of a randomly provisioned key and nonce. High-level abstractions should probably hide the nonce magic, or place nonce management in an opaque class exposed via an interface that a strong willed individual could implement if absolutely necessary.

**Comment by Drawaes** [comment]
@jnm2  That is the point they are all unmanaged memory situations (maybe except for the last one I haven't seen that before) the unsafe is you a releasing native resources or getting references to memory that the GC isn't going to help with and other potential effects.

Anyway I think that this is diverging, getting into naming at this point. It's more about the API shape, naming comes later.

@mmaguigan That is kinda what I am saying but you said it better, it should just "do the right defaults" but you can provide it the stuff you want.

**Comment by vcsjones** [comment]
>(maybe except for the last one I haven't seen that before)

I think that comes from the mono world.

While I whole heartedly agree with pointing things to safer than "primitive" primitives, I think it's a bigger discussion and this issue should be scoped to an implementation of AES-GCM that is "raw".

I would eventually like to broaden the subject of safety into something that works for everything, not just AES-GCM. Using AES-CBC without a MAC is dangerous, too, and in some circumstances CBC may be the correct thing to use instead of GCM. I would sooner see a safe (something like NaCl and cryptobox) API be a discussion that encompasses all of the crypto primitives, not just GCM.

**Comment by DemiMarie** [comment]
What about XChaCha20-Poly1305 with random 192-bit nonces?

**Comment by vcsjones** [comment]
>What about XChaCha20-Poly1305 with random 192-bit nonces?

That'll be a separate issue. People will still ask for AES-GCM. However, Windows CNG does not support any variant of ChaCha, and CoreFx relies on CNG for Windows cryptographic support. I don't believe OpenSSL supports XChaCha right now (Yes I am aware it supports ChaCha), either which is required for Linux support. The CoreFx team has a hard policy on not developing their own primitives, rather using the platforms built-in primitives (discussed earlier in this thread).

**Comment by mmaguigan** [comment]
@vcsjones, however, we should try and massage this API enough so that a user might implement support for his or her favorite AEAD algorithm without much difficulty. For example, if I chose to implement my primitives by p/invoking libsodium, it would be ideal that my implementation implement abstractions provided by this API surface. This will give users choice in deciding the backing library they wish to use (eg. swap out OpenSSL with LibSodium, even allowing hardware abstractions an easy route into the framework 😄 without all of the craziness we have today in that space).

**Comment by sdrapkin** [comment]
DJB-based crypto (xChaCha20/xSalsa20 + Poly1305, nacl & derivatives, etc) provide AE but not AEAD. Many people forget that. Is there a standard for how DJB-based AE modes should incorporate "Additional Data" to achieve AEAD? (I'm not aware of such).

It is critical for .NET to decide whether to provide an AE interface, an AEAD interface, or both.

Even within AEAD certain nuanced yet important differences are possible unless nailed down. Two alternative implementations:
1. AD is validated as part of tag validation, but knowledge/possession of AD is not needed to decrypt (ie. if AD is lost but a valid key is known, the decryption will succeed if the tag-check-failure is ignored).
2.  AD participates in internal derivation of the enc/dec key, so any tampering/loss of AD makes decryption impossible (ex. [Inferno](http://securitydriven.net/inferno/) uses that stronger model).

**Comment by Drawaes** [comment]
I have forked this API discussion due to the large number of comments and the API design getting lost. I have proposed a concrete API design at #23629

**Comment by danmosemsft** [comment]
@bartonjs estimates 4 weeks remaining

**Comment by pgolebiowski** [comment]
Why moved from 2.1.0 to future? In August, last year, @karelz said:

> The priority is pretty high on our Security roadmap - see dotnet/designs#11.
We currently plan to land this feature in 2.1 (with 95% confidence). See official release dates for more details about timelines.

AES-GCM support is important.

**Comment by bartonjs** [comment]
@pgolebiowski Normal software schedule things: overruns, interrupts, and a bit of priority juggling.

When we started buttoning up for preview1 all of the crypto features which didn't have proposed-and-reviewed API got moved to the next version.  (Similar things happened for other areas, but I don't know how they prioritized when they hit the `[estimate of work remaining] > [time left in release]` problem.)

As for "Future" versus something named "2.2", I think that's just "we need to focus on release quality" and the psychological idea of "don't give it a name yet, because if it has a name people will be tempted to work on it instead of finishing this release".

**Comment by karelz** [comment]
Re: 2.2 vs. Future - the more important reason is "People will start asking why x is in Future and not in 2.2". That means, once we create 2.2 milestone (we plan to do it tomorrow BTW), area owners should in theory do a pass through bugs and mark important ones for 2.2 to prevent confusion all around.

@joshfree @bartonjs you might want to update the Security roadmap & give everyone heads up about what made it into 2.1 vs. not.

**Comment by galich** [comment]
No worries, we have been waiting for quite some time.


@pgolebiowski we started to use openssl/libressl for linux builds and using BCryptEncrypt on Windows. Our users don't have to wait ;)

**Comment by Blackbaud-PeteLukich** [comment]
@karelz i'm looking for the latest version of `AuthenticatedAesCng`, which [ShawnFa's 2009 blog post](https://blogs.msdn.microsoft.com/shawnfa/2009/03/17/authenticated-symmetric-encryption-in-net/) describes getting from here:

> Security.Cryptography.dll 1.3 or higher (from http://codeplex.com/clrsecurity)

Is there a NuGet package that contains this class?

**Comment by karelz** [comment]
I don't think the class is anywhere on nuget. cc @bartonjs @GrabYourPitchforks

**Comment by bartonjs** [comment]
@Blackbaud-PeteLukich Nope.  And the clrsecurity project hasn't been maintained in 3 years.

**Comment by galich** [comment]
@bartonjs what are .NET Core alternatives to non-maintained dated clrsecurity project?

**Comment by ektrah** [comment]
FWIW, [NSec](https://www.nuget.org/packages/NSec.Cryptography/) provides AES256-GCM for .NET Core based on libsodium.

**Comment by pgolebiowski** [comment]
how about having the crypto in .NET Core based on libsodium? why not base the code on the implementation of NSec? looks really good.

@ektrah @bartonjs

**Comment by AdamWillden** [comment]
I've swapped to using NSec since the recommendation. The only annoyance is the fact I'm having to put a libsodium DLL in my project to copy to output. Seems that, for whatever reason, it's difficult to configure the nuget package to pull it in. The term used might have been 'shadow dependency'. The only way it'd work as an official approach is if that were fixed.

**Comment by blowdart** [comment]
This thread has unfortunately gone off its original topic. A decision has been made on AES-GCM, hence I'm locking the thread.

**Comment by bartonjs** [comment]
Fixed via #31389.
